packages/core/src/utils/NumFuzz.ts

```diff
@@ -0,0 +1,417 @@
+ * TODO: Consier moving this to shapes/Shapes
+ *
+ * @param shapeType Name of shape type
+ * @returns shape maker function
+ */
+export const getShapeMaker = (
+  shapeType:
+    | "Circle"
+    | "Ellipse"
+    | "Equation"
+    | "Image"
+    | "Line"
+    | "Path"
+    | "Polygon"
+    | "Polyline"
+    | "Rectangle"
+    | "Text"
+): ShapeMakerFn => {
+  return {
+    Circle: makeCircle,
+    Ellipse: makeEllipse,
+    Equation: makeEquation,
+    Image: makeImage,
+    Line: makeLine,
+    Path: makePath,
+    Polygon: makePolygon,
+    Polyline: makePolyline,
+    Rectangle: makeRectangle,
+    Text: makeText,
+  }[shapeType];
+};
+
+/**
+ * Describes the common shape of all Shape maker functions
+ * TODO: Consier moving this to shapes/Shapes
+ */
+export type ShapeMakerFn = (
+  context: Context,
+  canvas: Canvas,
+  properties: Properties
+) => Shape;
+
+// ------------------------------ Function Definitions ------------------------------ //
+
+/**
+ * Describes a Penrose computation function
+ */
+export type FnDef = {
+  fnName: string;
+  shapeType?:
+    | "Circle"
+    | "Ellipse"
+    | "Equation"
+    | "Image"
+    | "Line"
+    | "Path"
+    | "Polygon"
+    | "Polyline"
+    | "Rectangle"
+    | "Text";
+  args: FnArg[];
+};
+
+/**
+ * Describes an argument to a Penrose computation function
+ */
+export type FnArg = FnContext | FnVector | FnNum | FnVectorRand | FnNumRand;
+
+/**
+ * Base argument properties
+ */
+type FnBase = {
+  name: string; // Name of the argument
+  shapeProp?: boolean; // Whether this argument is a shape property
+};
+
+/**
+ * Non-randomized VectorV argument
+ */
+export type FnVector = FnBase & {
+  tag: "VectorV";
+  contents: number[]; //fnNum[];
+};
+
+/**
+ * Randomized VectorV argument
+ */
+export type FnVectorRand = FnBase & {
+  tag: "VectorRand";
+  contents: FnNumRand[];
+};
+
+/**
+ * Context argument (placeholder)
+ */
+export type FnContext = {
+  tag: "Context";
+  name: string;
+};
+
+/**
+ * Non-randomized Number argument
+ */
+export type FnNum = FnBase & {
+  tag: "FloatV";
+  contents: number;
+};
+
+/**
+ * Randomized Number argument
+ */
+export type FnNumRand = FnBase & {
+  tag: "FloatRand";
+  min: number;
+  max: number;
+};
+
+// ---------------------------------- Fuzzer Types ---------------------------------- //
+
+/**
+ * Describes the results of a Fuzzing test: input, output, and whether the test passed
+ */
+export type FuzzTestResult = {
+  input: number[]; // input to function
+  output: ad.Outputs<number>; // output from function
+  passed: boolean; // true if output matches oracle; false, otherwise
+};
+
+/**
+ * Fuzzing options for a given computation function
+ */
+export type FuzzOptions = {
+  fnCfg: FnDef; // Function configuration
+  variation?: string; // Variation / seed (optional)
+  graphFn: typeof primaryGraph; // Function for generating the computation graph. TODO: Should also accept secondaryGraph
```

Comment:
I haven't read all the code in this file, but I'd guess it's best to just use `makeGraph` directly here, rather than case-splitting on `primaryGraph` vs `secondaryGraph`.